packages/arb-bridge-eth/contracts/rollup/Rollup.sol

```diff
@@ -0,0 +1,699 @@
+        require(node2.stakers(stakers[1]), "STAKER2_NOT_STAKED");
+
+        require(
+            node1.challengeHash() ==
+                ChallengeLib.challengeRootHash(
+                    nodeFields[0],
+                    nodeFields[1],
+                    nodeFields[2],
+                    executionCheckTime
+                ),
+            "CHAL_HASH"
+        );
+
+        // Start a challenge between staker1 and staker2. Staker1 will defend the correctness of node1, and staker2 will challenge it.
+        address challengeAddress =
+            challengeFactory.createChallenge(
+                address(this),
+                nodeFields[0],
+                nodeFields[1],
+                nodeFields[2],
+                executionCheckTime,
+                stakers[0],
+                stakers[1],
+                challengePeriodBlocks
+            );
+
+        challengeStarted(stakers[0], stakers[1], challengeAddress);
+
+        emit RollupChallengeStarted(challengeAddress, stakers[0], stakers[1], nodeNums[0]);
+    }
+
+    /**
+     * @notice Inform the rollup that the challenge between the given stakers is completed
+     * @dev completeChallenge isn't pausable since in flight challenges should be allowed to complete or else they could be forced to timeout
+     * @param winningStaker Address of the winning staker
+     * @param losingStaker Address of the losing staker
+     */
+    function completeChallenge(address winningStaker, address losingStaker) external override {
+        // Only the challenge contract can declare winners and losers
+        require(msg.sender == inChallenge(winningStaker, losingStaker), "WRONG_SENDER");
+
+        uint256 loserStake = amountStaked(losingStaker);
+        uint256 winnerStake = amountStaked(winningStaker);
+        if (loserStake > winnerStake) {
+            loserStake -= reduceStakeTo(losingStaker, winnerStake);
+        }
+
+        uint256 amountWon = loserStake / 2;
+        increaseStakeBy(winningStaker, amountWon);
+        loserStake -= amountWon;
+        clearChallenge(winningStaker);
+
+        // TODO: deposit extra loserStake into ArbOS
+        turnIntoZombie(losingStaker);
+    }
+
+    /**
+     * @notice Remove the given zombie from nodes it is staked on, moving backwords from the latest node it is staked on
+     * @param zombieNum Index of the zombie to remove
+     * @param maxNodes Maximum number of nodes to remove the zombie from (to limit the cost of this transaction)
+     */
+    function removeZombie(uint256 zombieNum, uint256 maxNodes) external whenNotPaused {
+        require(zombieNum <= zombieCount(), "NO_SUCH_ZOMBIE");
+        address zombieStakerAddress = zombieAddress(zombieNum);
+        uint256 latestStakedNode = zombieLatestStakedNode(zombieNum);
+        uint256 nodesRemoved = 0;
+        uint256 firstUnresolved = firstUnresolvedNode();
+        while (latestStakedNode > firstUnresolved && nodesRemoved < maxNodes) {
```

Comment:
I think this should be `latestStakedNode >= firstUnresolved && ...`

Reply:
Great catch, fixed